crates/egui/src/widgets/image.rs

```diff
@@ -43,15 +44,93 @@ use epaint::{util::FloatOrd, RectShape};
 #[must_use = "You should put this widget in an ui with `ui.add(widget);`"]
 #[derive(Debug, Clone)]
 pub struct Image<'a> {
-    source: ImageSource<'a>,
+    sources: ImageSources<'a>,
     texture_options: TextureOptions,
     image_options: ImageOptions,
     sense: Sense,
     size: ImageSize,
     pub(crate) show_loading_spinner: Option<bool>,
 }
 
+#[derive(Debug, Clone)]
+/// This type tells the [`Ui`] if the image is an animated Image and when to render the next frame.
+///
+/// This is used by [`Image::new_animation`]
+pub enum ImageSources<'a> {
+    Multi(Vec<(ImageSource<'a>, Duration)>),
+    Single(ImageSource<'a>),
+}
+
+impl<'a> ImageSources<'a> {
+    pub fn source(&self, ctx: &Context) -> &'a ImageSource<'_> {
+        match self {
+            ImageSources::Multi(imgs) => {
+                let now = (ctx.input(|i| i.time) * 1000.) as u128;
+                let frames: u128 = imgs.iter().map(|v| v.1.as_millis()).sum();
+                let pos = now % frames;
+                let mut cumulative_duration = 0;
+                let mut index = 0;
+                for (i, (_, duration)) in imgs.iter().enumerate() {
+                    cumulative_duration += duration.as_millis();
+                    if cumulative_duration >= pos {
+                        index = i;
+                        break;
+                    }
+                }
+                let (img, duration) = &imgs[index];
+                ctx.request_repaint_after(*duration);
+                img
+            }
+            ImageSources::Single(img) => img,
+        }
+    }
+}
+
+impl<'a> From<ImageSource<'a>> for ImageSources<'a> {
+    fn from(value: ImageSource<'a>) -> Self {
+        Self::Single(value)
+    }
+}
+
+impl<'a> From<Vec<(ImageSource<'a>, Duration)>> for ImageSources<'a> {
+    fn from(value: Vec<(ImageSource<'a>, Duration)>) -> Self {
```

Comment:
it could be helpful with a `struct AnimatedImageSource<'a>(Vec<(ImageSource<'a>, Duration)>)` to avoid complex types like this inside `ImageSources`